genesis/engine/coupler.py

```diff
@@ -581,3 +587,905 @@ def couple_grad(self, f):
+                    + fem_solver.elements_el_hessian[i_b, i, 1, i_e] @ p9[3:6]
+                    + fem_solver.elements_el_hessian[i_b, i, 2, i_e] @ p9[6:9]
+                )
+
+            energy[i_b] += 0.5 * V_dt2 * p9.dot(H9_p9) * damping_beta_factor
+
+        # Contact
+        self.compute_contact_energy(f)
+        for i_c in range(self.n_fem_floor_contact_pairs[None]):
+            pair = self.fem_floor_contact_pairs[i_c]
+            i_b = pair.batch_idx
+            if not self.batch_linesearch_active[i_b] or pair.active == 0:
+                continue
+            energy[i_b] += pair.energy
+
+    @ti.kernel
+    def init_linesearch(self, f: ti.i32):
+        fem_solver = self.fem_solver
+        dt = ti.static(self.sim._substep_dt)
+        dt2 = dt**2
+        for i_b in ti.ndrange(self._B):
+            self.batch_linesearch_active[i_b] = self.batch_active[i_b]
+            if not self.batch_linesearch_active[i_b]:
+                continue
+            self.linesearch_state[i_b].step_size = 1.0 / ti.static(self._linesearch_tau)
+            self.linesearch_state[i_b].m = 0.0
+
+        # x_prev, m
+        for i_b, i_v in ti.ndrange(self._B, fem_solver.n_vertices):
+            if not self.batch_linesearch_active[i_b]:
+                continue
+            self.linesearch_state_v[i_b, i_v].x_prev = self.v[i_b, i_v]
+            self.linesearch_state[i_b].m += self.pcg_state_v[i_b, i_v].x.dot(self.gradient[i_b, i_v])
+
+        self.compute_total_energy(f, self.linesearch_state.prev_energy)
+
+    @ti.kernel
+    def one_linesearch_iter(self, f: ti.i32):
+        fem_solver = self.fem_solver
+
+        # update vel
+        for i_b, i_v in ti.ndrange(self._B, fem_solver.n_vertices):
+            if not self.batch_linesearch_active[i_b]:
+                continue
+            self.v[i_b, i_v] = (
+                self.linesearch_state_v[i_b, i_v].x_prev
+                + self.linesearch_state[i_b].step_size * self.pcg_state_v[i_b, i_v].x
+            )
+
+        self.compute_total_energy(f, self.linesearch_state.energy)
+
+        # check condition
+        for i_b in ti.ndrange(self._B):
+            if not self.batch_linesearch_active[i_b]:
+                continue
+            self.batch_linesearch_active[i_b] = (
+                self.linesearch_state[i_b].energy
+                > self.linesearch_state[i_b].prev_energy
+                + self._linesearch_c * self.linesearch_state[i_b].step_size * self.linesearch_state[i_b].m
+            )
+            if not self.batch_linesearch_active[i_b]:
+                continue
+            self.linesearch_state[i_b].step_size *= self._linesearch_tau
+
+    def linesearch(self, f: ti.i32):
+        """
+        Note
+        ------
+        https://en.wikipedia.org/wiki/Backtracking_line_search#Algorithm
+        """
+        self.init_linesearch(f)
+        for i in range(self._n_linesearch_iterations):
+            self.one_linesearch_iter(f)
```

Comment:
I wonder if we can use a single kernel launch for this, i.e. put the iteration loop inside the kernel.

Reply:
we may think about other way in the future. I think this one is good for now.